car/README.md

```diff
@@ -0,0 +1,108 @@
+# Certified ARchive
+
+CARs (Certified ARchives) are archives for IPLD DAGs.
+
+## Summary
+
+CARs are archives of IPLD DAGs. They are:
+
+1. Certified
+2. Seekable
+3. Compact
+4. Reproducible
+5. Simple and Stable
+
+The actual format is just a (mostly) recursively-defined topological sort of the
+desired DAG with some metadata for fast traversal.
+
+```
+CID
+len(ROOT)
+ROOT
+[
+  CHILD-1-OFFSET
+  ...
+]
+[
+  len(CHILD-1)
+  CHILD-1
+  [
+    CHILD-1/1-OFFSET
+    ...
+  ]
+  ...
+]
+```
+
+Offsets are relative, offsets for missing children use a sentinel value.
```

Comment:
So how would you deal with a leaf node that is referenced from multiple branches, e.g. root -> a -> c, root -> b -> c?

I assume that you would store c only once and just calculate the relative offset for both a and b? In that case a and b would just have offset arrays, but not value arrays, and c would be stored at the root level? Would c then be stored before a and b?

Reply:
>  assume that you would store c only once and just calculate the relative offset for both a and b? In that case a and b would just have offset arrays, but not value arrays, and c would be stored at the root level?

Yes. I'll try to make this diagram a bit less confusing (and add the example below).

> Would c then be stored before a and b?

That's where the topological sort comes in. We'd store this as:

```
      root.Cid()
root: root.Length()
      root.Bytes()
      offset(a)
      offset(b)
a:    a.Length()
      a.Bytes()
      offset(c)
b:    b.Length()
      b.Bytes()
      offset(c) // 0
c:    c.Length()
      c.Bytes()
```

Reply:
Well, breadth first may not be the best method to keep children close to there parents, in fast there probably is no best method to do so.  Nevertheless, a still think it would be better to allow negative offsets to allow children to be close to there first parent when possible.

Reply:
> This could force certain children to be very far from there first parent and will likely not stream well.

We can't both stream (write) and support the offset index (unless we keep the index in memory and append it to the end but that would hurt reading significantly).

> children close to there (first) parent

Regardless of what we do, we can't get this property (assuming a reasonable branching factor). For example, with a branching factor of 2, some children will already be ~1024 nodes away from their parents once we hit a depth of 10.

However, this would ensure that siblings are (usually) next to each other. On the other hand, we could probably get a similar property with a topological sort if we use the right algorithm (although we'd always end up grouping nodes with the deepest siblings).

> A negative value can then be meant as a clear indicator of a duplicate, and with some luck the node may already be in the cache and it will not be necessary to seek in order to gets the contents.

We'd have to have read in the entire CAR for this to be the case. It would be nice to be able to skip over portions of the CAR that we don't care about.

---

Basically, if we can't optimize for streaming while writing, we might as well optimize for streaming while reading. With a topologically sorted DAG, we can traverse a DAG in one pass through the file by:

1. Looking a at a node.
2. Picking the child we want.
3. Looking up the offset of that child.
4. Seeking (forward) to the offset of that child.
5. Recurse.

This is a useful property for media that works best with sequential access patterns.

Reply:
>  Nevertheless, a still think it would be better to allow negative offsets to allow children to be close to there first parent when possible.

Personally, I don't really see any benefit of putting them close to the first parent rather than the last. I don't see order as having anything to do with "importance". Is there any reason *to* do this?